src/languages/epp.js

```diff
@@ -0,0 +1,32 @@
+/*
+Language: EPP (Embedded Puppet templating language)
+Requires: xml.js, puppet.js
+Author: Ben Ford <ben.ford@puppet.com>
+Contributors: Antoine Musso <hashar@free.fr>
+Description: "Bridge" language defining fragments of Puppet in plain text within <% .. %>
+Website: https://www.puppet.com/docs/puppet/8/lang_template_epp.html
+Category: template
+*/
+
+/** @type LanguageFn */
+export default function(hljs) {
+  return {
+    name: 'EPP',
+
+    // The actual content can be any type
+    subLanguage: [],
```

Comment:
This is problematic as it relies on auto-detect which has fallen a bit out of favor (and honestly has never worked super well)... that's going to make the `anysublanguage.txt` test fragile, as well as working poorly in general cases.  Is there a MUCH shorter list of common embedded environments we could list explicitly?

I don't think we're interested in saying:

> Just use "epp" and we'll figure it out.

When we know that's not really true and will only lead to incorrect highlighting and support requests.

What you'd really want (I think) is a way to specify you want `epp-xml` and then handle just that case... But I'm not sure how you could specify that in a CSS name dynamically without it being confused as the grammar `epp-xml` rather than `epp (embedded within XML)`.

Thoughts?

Reply:
The "problem" with those templating languages (epp, erb, jinja from Django) is they can be used to represent anything though maybe it is not going to be used to generate a compiled language source code (`csharp`, `c`, `delphi` etc) there is no real limit as to which target format it can represent (our repo has Apache config, bash, js, json, lua, sql, php, python, ruby, systemd, toml, plain text, xml, yaml to name a fews).

I wrote the first markup test with something looking like an `ini` file which was indeed erroneously detected (as `csharp`, potentially due to a `<% # this is a comment %>`). Retrying today with `test/markup/epp/anysublanguage.txt`:
```
<% # this is a comment %>
[section]
foo = <%= ec2_metadata["ami_id"] %>
```
Is autodetected as `erlang` so yeah that is really fragile hence why I went with a test using XML which is properly auto detected. But autodetections kick in on each block and `test/markup/epp/anysublanguage.expect.txt` has a block for `mercury`:

    <span class="language-mercury"><span class="hljs-comment">%&gt;&lt;/ami-id&gt;</span>

So that is clearly wrong and **`subLanguage: []` is going to be nightmare** and it should be removed. So if one process an `epp` file they will have solely the templating markers and embedded Puppet rendered but the rest of the file untouched (plain text?).

----

From the discussion I had with one of my colleague on that topic, I mentioned vim and its support to mix up syntax for a file. An erb template representing a javascript file can be held in a file like `template.js.erb` and one can hint vim about it with: 

    autocmd BuffRead *.js.erb set ft=eruby.javascript

Which tells vim to apply erb first then javascript (a screenshot for an erb template generating a ini file: https://phabricator.wikimedia.org/F37104450 ).

> What you'd really want (I think) is a way to specify you want `epp-xml` and then handle just that case... But I'm not sure how you could specify that in a CSS class name dynamically without it being confused as the grammar `epp-xml` rather than `epp` (embedded within XML).

The different languages have their own `<span>` with the CSS class for the language (`language-xml` or `language-puppet`). From `test/markup/epp/anysublanguage.expect.txt`:
```
<span class="language-xml">&lt;?xml version=<span class="hljs-string">&quot;1.0&quot;</span>
...
<span class="language-puppet"><span class="hljs-variable">$ec2_metadata</span></span>
```

Can the sublanguages be given when invoking highlightjs? If the caller knows the file is an erb for a javascript (maybe cause of the convention the filename has a suffix of `.js.erb`) it could invoke highlightjs with the sublanguage set explicitly, something like: `highlightjs(language='erb, sublanguage['js'])`.

I will rework the pull request to remove the `subLanguage: []`.

Reply:
> the rest of the file untouched (plain text?).

Well, it's not `subLanguage: "plaintext"` exactly but it's essentially equivalent since our plain text grammar does nothing.

> Can the sublanguages be given when invoking highlightjs? 

I was worried more about use in HTML, ie `<pre><code class="lang-x">` style usage, which I consider the harder question, how does one specify arguments here.  I dunno about "when invoking"... grammars are "compiled" at first use... they aren't things that are very dynamic  - and certainly there currently exists no way to pass arguments to a grammar.  That said writing an extension that did this would be trivial... something along the lines of:

```
// get the compiled grammar
var epp = hljs.getLanguage("epp")
// get the raw source of the grammar before compilation
var epp_xml = l.rawDefinition;
// alter the grammar to specify XML
epp_xml.subLanguages= ["xml"]
// re-register and recompile your new grammar
hljs.registerLanguage("epp-xml", epp_xml)
// use your new grammar to highlight code
hljs.highlight(code, {language: "epp-xml"})
```

To me someone already using the API explicitly (vs just auto mode) to highlight blocks would just do something along these lines. I'm not yet seeing an approachable simple way of adding this directly to core.

Reply:
One could then imagine an API that took a function that did a dynamic transform of a specified grammar:

```js
hljs.highlightTransform(code, {language: "epp"}, (def) => { def.subLanguage = ["xml"] })
```

We could even ship a "boring" version of "epp" in core and then you could publish a short wrapper on the Wiki that let someone choose the sublanguage of their choice if they were willing to add a little custom JS to their project.